### ISSUE TITLE
predict.dfm

### ISSUE THREAD

**Issue by zachmayer** [kbenoit/quanteda#46]
It would be really useful to be able project new texts into the same feature space as an existing `dfm`.  This would be particularly useful if you're using texts as inputs to a predictive model.

**Comment by kbenoit** [comment]
We've created a new `selectFeatures()` method for dfm objects, so you should now be able to do:
```{r}
dfm1 <- dfm(inaugTexts[1:3])
dfm2 <- dfm(inaugTexts[4:5])
dfm2b <- selectFeatures(dfm2, features(dfm1))
```

Is that what you meant?

**Comment by zachmayer** [comment]
I'm interested in using text data for predictive modeling, where the test-set texts may not be known at training time.  Here's an example of what I'd like to do:

**Train a model:**
```{R}
library(quanteda)
library(glmnet)
target <- as.integer(inaugCorpus$documents$President == 'Roosevelt')
train <- c(18, 26, 33, 48, 37, 38, 39, 40, 53, 55)
test <- c(1,  3,  4, 14, 30, 51)

#Fit a model
train_tokens <- dfm(
  inaugTexts[train],
  toLower = TRUE,
  stem = TRUE,
  removeTwitter = FALSE,
  removeNumbers = FALSE,
  removePunct = TRUE,
  removeSeparators = TRUE,
  what = 'word',
  ngrams = 1:3
  )
model <- glmnet(train_tokens, target[train], family='binomial')
CF <- coef(model)[,100]
sort(CF[CF!=0])
```

So far so good!

Now I want to get predictions for new data, which brings up 3 problems:
1. I need to remember the settings I passed to the training set dfm.  This isn't too bad to do, but it'd sure be convenient to include them with the `dfmSparse` in case I forget, or pass the model on to someone else.
2. I need to remove words in the test set that are not in the training set. (Your new function does this just fine).
3. I need to add dummy columns with all 0's for words in the training set that are not in the test set.
Ideally, a single `predict` function would take care of all 3 steps, generating a `dfmSparse` for *new* data that has the exact same pre-processing and dimensionality as the data use to generate the original matrix:
```{R}
#Generate a new dfm matrix based on the specification of an old dfm matrix
test_tokens <- predict(train_tokens, inaugTexts[test])
probs <- predict(model, test_tokens)
```

**Comment by zachmayer** [comment]
Here is my dirty hack for predicting on new data:
```{R}
#Make predictions for new data
test_tokens <- dfm(
  inaugTexts[test],
  toLower = dfm_args$toLower,
  stem = dfm_args$stem,
  removeTwitter = dfm_args$removeTwitter,
  removeNumbers = dfm_args$removeNumbers,
  removePunct = dfm_args$removePunct,
  removeSeparators = dfm_args$removeSeparators,
  what = dfm_args$what,
  ngrams = dfm_args$ngrams
)
test_tokens <- selectFeatures(test_tokens, train_tokens) #Fails
test_tokens <- test_tokens[,intersect(colnames(test_tokens), colnames(train_tokens))]
for(col in setdiff(colnames(train_tokens), colnames(test_tokens))){
  test_tokens <- cBind(test_tokens, 0)
  colnames(test_tokens)[ncol(test_tokens)] <- col
}
test_tokens <- drop0(test_tokens)
test_tokens <- test_tokens[,colnames(train_tokens)]
probs <- predict(model, test_tokens, type='response')
probs[,100]
```

This would be a lot more elegant: :grin:
```{R}
test_tokens <- predict(train_tokens, inaugTexts[test])``` 
And probably more efficient too.

**Comment by tyokota** [comment]
I think you forgot to identify the features you want to intersect the two dfm objects with features()
   test_tokens <- selectFeatures(test_tokens, features(train_tokens))

Another solution I have used is as follows. This is pretty "rough", but using the docvars parameter in corpus() might help.

df.train$trainclass <- 1
df.test$trainclass  <- 2
df.full <- rbind(df.test, df.train)

corpus.full <- corpus(as.matrix(df.full$text),
                     docvars=data.frame(trainclass=as.matrix(df.full$trainclass)),
                     notes="A corpus")

dfm.full <- dfm(...)

dfm.train <- dfm.full[docvars(corpus.full, "trainclass")==1,]
dfm.test  <- dfm.full[docvars(corpus.full, "trainclass")==2,]

library(glmnet)
model <- glmnet(dfm.train, target[train], family='binomial')
CF <- coef(model)[,100]
sort(CF[CF!=0])

**Comment by zachmayer** [comment]
I'll test with `test_tokens <- selectFeatures(test_tokens, features(train_tokens))`.

My general problem is the test documents aren't usually know at the time I fit the model.  E.g. I want to put this model into production somewhere and score new documents as they come into our server.

**Comment by zachmayer** [comment]
So here's the problem: my training set is 6 x 34,117, so my test set needs to have that exact dimension:
```{R}
> dim(train_tokens)
[1]    10 34117
> test_tokens <- selectFeatures(test_tokens, features(train_tokens))
kept 4,662 features, from 34117 supplied feature types
> dim(test_tokens)
[1]    6 4662
```

My trainset has 34k columns but my test set has only 4.6k columns. So when I go to predict, I get an error:
```{R}
> probs <- predict(model, test_tokens, type='response')
 Hide Traceback
 
 Rerun with Debug
 Error in as.matrix(cbind2(1, newx) %*% nbeta) : 
  error in evaluating the argument 'x' in selecting a method for function 'as.matrix': Error in cbind2(1, newx) %*% nbeta : 
  Cholmod error 'A and B inner dimensions must match' at file ../MatrixOps/cholmod_ssmult.c, line 82 
5 as.matrix(cbind2(1, newx) %*% nbeta) 
4 predict.glmnet(model, test_tokens, type = "response") 
3 NextMethod("predict") 
2 predict.lognet(model, test_tokens, type = "response") 
1 predict(model, test_tokens, type = "response") 
```

Long story short, it would be really useful to be able to apply the text transformation from a given corpus to completely new, previously unseen data.  This would make predictive modeling a lot easier! :grin: 

For now, I'll use my dirty hack to add in all the missing columns, but I don't like it:
```{R}

test_tokens <- test_tokens[,intersect(colnames(test_tokens), colnames(train_tokens))]
for(col in setdiff(colnames(train_tokens), colnames(test_tokens))){
  test_tokens <- cBind(test_tokens, 0)
  colnames(test_tokens)[ncol(test_tokens)] <- col
}
```

**Comment by kbenoit** [comment]
Looks to me like your `predict()` command is failing because your test dfm has fewer features than the training dfm. I would suggest you create just one dfm consisting of both training and test documents, and then slice out the cases you want for training the model, and apply the same slice in negative to apply the prediction to the test set. This should ensure the same number of features without the "ugly hack". 

I'm happy to experiment if you can send some sample data illustrating the problem, so that I can replicate the problem.

**Comment by zachmayer** [comment]
>I would suggest you create just one dfm consisting of both training and test documents
Unfortunately, this isn't possible in my application.  I'm trying to score NEW documents that didn't even exist when I trained the model!

**Comment by kbenoit** [comment]
OK, I think I have addressed the original issues 2 & 3. (Issue 1 is different, since it has to do with adding settings to the dfm object. Primitive settings already exist, and adding more is on the development list. I will open a separate issue for this.)

Of course you will need to pull this from the master on GitHub, as it will be another week or so before it finds its way to the CRAN version.

I did not add a `predict` method since that has a special meaning for models, and would not apply to constructing matrix objects. But I think what I have done is just as convenient. I also optimised `selectFeatures()` (which is called by `dfm(x, keptFeatures = something)`) so that it is much faster now, unless you use `valuetype = "regex"`.

    > textVec1 <- c("This is text one.", "This, the second text.", "Here: the third text.")
    > textVec2 <- c("Here are new words.", "New words in this text.")
    > (dfm1 <- dfm(textVec1))
    Creating a dfm from a character vector ...
       ... lowercasing
       ... tokenizing
       ... indexing documents: 3 documents
       ... indexing features: 8 feature types
       ... created a 3 x 8 sparse dfm
       ... complete. 
    Elapsed time: 0.033 seconds.
    Document-feature matrix of: 3 documents, 8 features.
    3 x 8 sparse Matrix of class "dfmSparse"
           features
    docs    this is text one the second here third
      text1    1  1    1   1   0      0    0     0
      text2    1  0    1   0   1      1    0     0
      text3    0  0    1   0   1      0    1     1
    > (dfm2a <- dfm(textVec2))
    Creating a dfm from a character vector ...
       ... lowercasing
       ... tokenizing
       ... indexing documents: 2 documents
       ... indexing features: 7 feature types
       ... created a 2 x 7 sparse dfm
       ... complete. 
    Elapsed time: 0.007 seconds.
    Document-feature matrix of: 2 documents, 7 features.
    2 x 7 sparse Matrix of class "dfmSparse"
           features
    docs    here are new words in this text
      text1    1   1   1     1  0    0    0
      text2    0   0   1     1  1    1    1
    > (dfm2b <- dfm(textVec2, keptFeatures = dfm1))
    Creating a dfm from a character vector ...
       ... lowercasing
       ... tokenizing
       ... indexing documents: 2 documents
       ... indexing features: 7 feature types
       ... found 3 features from 8 supplied types in a dfm, padding 0s for another 5 
       ... created a 2 x 8 sparse dfm
       ... complete. 
    Elapsed time: 0.024 seconds.
    Document-feature matrix of: 2 documents, 8 features.
    2 x 8 sparse Matrix of class "dfmSparse"
          this is text one the second here third
    text1    0  0    0   0   0      0    1     0
    text2    1  0    1   0   0      0    0     0
    > identical(features(dfm1), features(dfm2b))
    [1] TRUE

**Comment by mlandry22-h2o** [comment]
Hopefully I don't re-open this, but I did want to leave some feedback, that this line:
```
dfm2b <- dfm(textVec2, keptFeatures = dfm1)
```
produced exactly what I needed. I think that's likely what Zach wanted as well, but can't say for sure. It's a really nice feature.

**Comment by kbenoit** [comment]
Thanks for the feedback! Glad this works for you. 👍
